File: src/main/java/org/javiermf/features/models/ProductConfiguration.java
package org.javiermf.features.models;

import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

@Entity
public class ProductConfiguration {

    @Id
    @GeneratedValue
    Long id;

    @Column(nullable = false)
    String name;

    @JsonIgnore
    @ManyToOne
    Product product;

    @ManyToMany(fetch = FetchType.EAGER)
    Set<Feature> activedFeatures = new HashSet<Feature>();

    public Set<Feature> getActivedFeatures() {
        return activedFeatures;
    }

    public Set<String> availableFeatures() {
        return collectFeatureNames(product.getProductFeatures());
    }

    private Set<String> collectFeatureNames(Set<Feature> featuresSet) {
        Set<String> features = new HashSet<String>();
        for (Feature feature : featuresSet) {
            features.add(feature.name);
        }

        return features;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Product getProduct() {
        return product;
    }

    public void setProduct(Product product) {
        this.product = product;
    }

    public Set<String> activedFeatures() {
        return collectFeatureNames(activedFeatures);
    }

    public void active(Feature feature) {
        activedFeatures.add(feature);

    }

    public void deactive(Feature feature) {
        activedFeatures.remove(feature);
    }

    public void active(String featureName) {
        for (Feature feature : product.getProductFeatures()) {
            if (featureName.equalsIgnoreCase(feature.getName())) {
                active(feature);
            }
        }

    }
}

File: src/main/kotlin/org/javiermf/features/models/ProductConfiguration.kt
package org.javiermf.features.models

import com.fasterxml.jackson.annotation.JsonIgnore
import java.util.*
import javax.persistence.*

@Entity
class ProductConfiguration {

    @Id
    @GeneratedValue
    internal var id: Long? = null

    @Column(nullable = false)
    var name: String

    @Column
    var isValid: Boolean? = true

    @JsonIgnore
    @ManyToOne
    var product: Product

    @ManyToMany(fetch = FetchType.EAGER)
    internal var activedFeatures: MutableSet<Feature> = HashSet()

    constructor(name: String, product: Product) {
        this.name = name
        this.product = product
    }

    fun getActivedFeatures(): Set<Feature> {
        return activedFeatures
    }

    fun availableFeatures(): Set<String> {
        return collectFeatureNames(product.getProductFeatures())
    }

    private fun collectFeatureNames(featuresSet: Set<Feature>): Set<String> {
        val features = HashSet<String>()
        for (feature in featuresSet) {
            features.add(feature.name)
        }

        return features
    }

    fun activedFeatures(): Set<String> {
        return collectFeatureNames(activedFeatures)
    }

    fun active(feature: Feature) {
        activedFeatures.add(feature)

    }

    fun deactive(feature: Feature) {
        activedFeatures.remove(feature)
    }

    fun active(featureName: String) {
        val feature = product.findProductFeatureByName(featureName)
        active(feature)
    }

    fun deactive(featureName: String) {
        val feature = product.findProductFeatureByName(featureName)
        deactive(feature)
    }

    fun hasActiveFeature(featureName: String): Boolean {
        return activedFeatures().contains(featureName)
    }
}
